File: solutions/c/collatz-conjecture/1/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start) {
   if (start <= 0) {
      return ERROR_VALUE;
   }

   int steps = 0;

   while (start > 1) {
      steps++;

      if (start % 2 == 0) {
         start /= 2;
      } else {
         start = start * 3 +1;
      }
   }

   return steps;
}

File: solutions/go/collatz-conjecture/1/collatz_conjecture.go
package collatzconjecture

import "fmt"

func CollatzConjecture(n int) (int, error) {
    if n < 1 {
        return 0, fmt.Errorf("Input has to be 1 or higher (%d given)", n)
    }

    var steps int
    
    for n > 1 {
        steps += 1
        
        if n % 2 == 0 {
            n = n / 2
        } else {
            n = n * 3 + 1
        }
    }

    return steps, nil
}
